crates/shared/src/order_quoting.rs

```diff
@@ -530,6 +548,78 @@ impl OrderQuoting for OrderQuoter {
     }
 }
 
+/// Handles whether or not quotes need to be verified.
+enum QuoteVerification {
+    /// Quotes get trusted blindly.
+    Never,
+    /// Enforce quotes only when the user has the required sell token balance.
+    /// If a user has the required sell token balance and a quote can still not
+    /// be verified the trade likely involves a weird token that solvers are not
+    /// able to handle correctly.
+    WhenFunded(Arc<dyn BalanceFetching>),
+}
+
+impl QuoteVerification {
+    /// Makes sure a quote was verified according to the configured rule.
+    async fn check(
+        &self,
+        estimate: &Estimate,
+        parameters: &QuoteParameters,
+        sell_amount: U256,
+    ) -> Result<(), CalculateQuoteError> {
+        let Some(verification) = &parameters.verification else {
+            return Ok(()); // no verification requested
+        };
+        if estimate.verified {
+            return Ok(()); // successfully verified
+        }
+        let balance_fetcher = match &self {
+            QuoteVerification::Never => return Ok(()),
+            QuoteVerification::WhenFunded(fetcher) => fetcher.as_ref(),
+        };
+
+        let balance =
+            match Self::get_balance(balance_fetcher, verification, parameters.sell_token).await {
+                Ok(balance) => balance,
+                Err(err) => {
+                    tracing::warn!(?err, "could not fetch balance for verification");
+                    return Err(CalculateQuoteError::QuoteNotVerified);
```

Comment:
I am still not very familiar with the way the errors are logged/forwarded to the elasticsearch. But wouldn't it make more sense to have different error types for this and the one in the [L593](https://github.com/cowprotocol/services/pull/2621/files#diff-3f8b6079ce50ae3ab0fe449d91c7e68922cae4023902961c892b58f91ddc49b3R593)? That way they would be logged differently and it would be easier to identify the causes if the error is triggered.

Reply:
These errors eventually get forwarded as an error response to whoever called the orderbook API. While technically these are different errors I think we should not expose too many details and group broad categories of errors together.
Essentially the user is the consumer of this particular error and for them it seems like the most important aspect is that the quote could not be verified and not that we had an temporary node issue or sth like that.

Reply:
@MartinquaXD got it! I fully agree with you about not exposing too many details. My point/concern was more on the logging side. If an user complains he/she got `QuoteNotVerified`, if we log both errors (this, and the one in L593) as `QuoteNotVerified`, wouldn't it be trickier for us to detect which one was? If we were to have two errors here, but then both derive to `QuoteNotVerified` as an external error, in the logs we would see the two different errors, one for each case (easier to debug/troubleshoot for us), but the user only sees `QuoteNotVerified` :thinking: I don't know if I am explaining myself and if it makes sense in our current system, I need to study more how the errors are logged, so I may be saying right now not applicable stuff.

Reply:
I don't think that adding a different error for something which is indeed a different kind of error is overengineering, But let's agree to disagree.